feat(mito2): WAL replay

v0y4g3r opened the pull request:
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR adds WAL replay support to mito2.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

evenyag commented on src/mito2/src/engine/tests.rs:
```diff
@@ -285,3 +285,63 @@ async fn test_write_to_region() {
+async fn test_region_replay() {
+    common_telemetry::init_default_ut_logging();
+    let env = TestEnv::with_prefix("region-replay");
+    let (engine, logstore, object_store) = env.create_engine(MitoConfig::default()).await;
```
Could we keep the logstore and object store in the `TestEnv`? Then we can get them via getters and don't need to return them in `create_engine`

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -16,41 +16,50 @@
     memtable_builder: MemtableBuilderRef,
     object_store: ObjectStore,
     region_dir: String,
+    wal: Wal<S>,
```
question: We only need wal when opening the region, could we pass it in `RegionOpener::open()`?
```rust
pub(crate) async fn open<S: LogStore>(self, config: &MitoConfig, wal: Wal<S>) -> Result<MitoRegion> {}
```

evenyag commented on src/mito2/src/region_writer.rs:
```diff
@@ -0,0 +1,169 @@
+use crate::wal::{EntryId, WalWriter};
+
+/// Context to keep region metadata and buffer write requests.
+///
```
```suggestion
```

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
+    flushed_sequence: SequenceNumber,
+    version_control: VersionControlRef,
+) -> Result<()> {
+    let mut notify_waiter = vec![];
```
We don't need to notify waiters during replay.

v0y4g3r replied:
> We don't need to notify waiters during replay.

but we have to wait until all entries' inserted to memtable during replay.

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
+) -> Result<()> {
+    let mut notify_waiter = vec![];
+    {
+        let mut region_writer_ctx = RegionWriteCtx::new(region_id, version_control);
```
```suggestion
        let mut region_write_ctx = RegionWriteCtx::new(region_id, version_control);
```

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
+    let mut notify_waiter = vec![];
+    {
+        let mut region_writer_ctx = RegionWriteCtx::new(region_id, version_control);
+        let mut wal_stream = wal.scan(region_id, flushed_sequence)?;
+        while let Some(res) = wal_stream.next().await {
+            let (_, entry) = res?;
```
We might need to store `flushed_entry_id` and use it instead of `flushed_sequence`.

We must also initialize `last_entry_id` and `committed_sequence` to update them to correct values after `write_memtables()`. I updated the entry id and committed sequence in https://github.com/GreptimeTeam/greptimedb/pull/2230

v0y4g3r replied:
Maybe add a todo here?

evenyag replied:
Maybe a FIXME

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
+        let mut wal_stream = wal.scan(region_id, flushed_sequence)?;
+        while let Some(res) = wal_stream.next().await {
+            let (_, entry) = res?;
+            for mutation in entry.mutations.into_iter() {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                region_writer_ctx.push_mutation(mutation.op_type, mutation.rows, Some(tx));
+                notify_waiter.push(rx);
+            }
```
1. We don't need to reply response to users so there is no need to create a channel.
2. Do we need `into_iter()`?
```suggestion
            for mutation in entry.mutations {
                region_writer_ctx.push_mutation(mutation.op_type, mutation.rows, None);
            }
```

evenyag commented on src/mito2/src/region/opener.rs:
```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
+        region_writer_ctx.write_memtable();
+    }
+
+    let rows_replayed: usize = futures::future::try_join_all(notify_waiter)
+        .await
+        .context(RecvSnafu)?
+        .into_iter()
+        .collect::<Result<Vec<Output>>>()?
+        .into_iter()
+        .map(|output| {
+            let Output::AffectedRows(rows_inserted) = output else {
+                unreachable!("Only expect affected rows during WAL replay.")
+            };
+            rows_inserted
+        })
+        .sum();
```
We can sum the inserted rows in the replay loop. Then receiver of the write notify is used to notify callers outside.

evenyag commented on src/mito2/src/region_writer.rs:
Should we rename it to `region_write_ctx.rs`? Or move it to `src/region/write_ctx.rs`?

evenyag commented on src/mito2/src/region_writer.rs:
```diff
@@ -0,0 +1,169 @@
+
+impl RegionWriteCtx {
+    /// Returns an empty context.
+    pub(crate) fn new(region_id: RegionId, version_control: VersionControlRef) -> RegionWriteCtx {
```
non-blocking: We can pass version_control by ref since we only need to call `current()`.
```suggestion
    pub(crate) fn new(region_id: RegionId, version_control: &VersionControlRef) -> RegionWriteCtx {
```

evenyag commented on src/mito2/src/wal.rs:
```diff
@@ -51,6 +51,12 @@ impl<S> Wal<S> {
         Self { store }
     }
+
+    pub fn cloned(&self) -> Wal<S> {
+        Wal {
+            store: self.store.clone(),
+        }
+    }
```
Why do we need this?

evenyag commented on src/mito2/src/worker.rs:
```diff
@@ -367,7 +367,7 @@ impl<S: LogStore> RegionWorkerLoop<S> {
 }
 
-impl<S> RegionWorkerLoop<S> {
+impl<S: LogStore> RegionWorkerLoop<S> {
```
Since we need `S: LogStore`, we can move this method to the `impl block` above.
```rust
impl<S: LogStore> RegionWorkerLoop<S> {
    async fn handle_requests(&mut self, buffer: &mut RequestBuffer) {}

    async fn handle_ddl_requests(&mut self, ddl_tasks: Vec<RegionTask>) {}
}

impl<S> RegionWorkerLoop<S> {
    async fn clean(&self) {}
}
```

v0y4g3r commented:
@evenyag PTAL. I added support for updating last entry id during replay and added test that asserts if `last_entry_id` and `committed_sequence` are correct after replay.

evenyag commented on src/mito2/src/engine/tests.rs:
```diff
@@ -287,11 +288,100 @@ async fn test_write_to_region() {
+        .handle_request(region_id, RegionRequest::Put(RegionPutRequest { rows }))
+        .await
+        .unwrap();
+    let Output::AffectedRows(rows_inserted) = output else {
+        unreachable!()
+    };
+    assert_eq!(20, rows_inserted);
```
We already test output in write tests, we don't need to assert it in all remaining tests.

waynexia reviewed (approved):
LGTM
